### PR TITLE
fix assets with the same backfill policy not being materialized together

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
A user reported an issue where a backfill requested a parent and child asset partitions in separate runs, causing the child asset partition to fail since the parent had not yet been materialized. After some investigation, it looks like if multiple assets have the same partitions definition, different backfill policies, and were going to be requested in one tick of the backfill daemon, we requested each asset individually instead of grouping by backfill policy. This causes parent-child asset pairs that were going to be materialized together by that tick to be requested in different runs.

This PR changes the behavior so that we group by backfill policy. I think this change should be ok because we already know the assets have the same partitions def and requested partitions, and we know that any parent/child assets must have the same backfill definition from the `can_run_with_parent` check earlier in the daemon loop

## How I Tested These Changes
new unit test. It failed before making the code changed (there were 3 run requests instead of 2) and passes now

### CODE REVIEW

**Comment by jamiedemaria** [comment]
* **#21880** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21880?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @jamiedemaria and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by jamiedemaria** [comment]
@sryza since Claire is out until tomorrow, could you take a look at this? I'd like to get the fix into this week's release if possible

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py [hunk 1]
```diff
@@ -485,37 +485,26 @@ def build_run_requests_with_backfill_policies(
...
-            if len(backfill_policies) == 1:
-                # if all backfill policies are the same, we can backfill them together
-                backfill_policy = backfill_policies.pop()
+            asset_keys_by_backfill_policy = defaultdict(set)
+            for asset_key in asset_keys:
+                asset_keys_by_backfill_policy[asset_graph.get(asset_key).backfill_policy].add(
+                    asset_key
+                )
+
+            for backfill_policy, aks in asset_keys_by_backfill_policy.items():
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:494]: Would it make sense to do all the grouping in one place? I.e. `assets_to_reconcile_by_partitions_def_partition_keys_backfill_policy`?

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py]: yeah good point - done!

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py [hunk 2]
```diff
@@ -324,21 +324,24 @@ def build_run_requests_with_backfill_policies(
...
             asset_partition_keys[asset_partition.asset_key].add(asset_partition.partition_key)
 
-    assets_to_reconcile_by_partitions_def_partition_keys: Mapping[
-        Tuple[Optional[PartitionsDefinition], Optional[FrozenSet[str]]], Set[AssetKey]
+    assets_to_reconcile_by_partitions_def_partition_keys_backfill_policy: Mapping[
+        Tuple[Optional[PartitionsDefinition], Optional[FrozenSet[str]], Optional[BackfillPolicy]],
+        Set[AssetKey],
     ] = defaultdict(set)
 
     # here we are grouping assets by their partitions def and partition keys selected.
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:332]: Update this comment or remove it?

**Review by sryza — APPROVED** [review]

**Pull request merged**